src/osd/scrubber/osd_scrub.h

```diff
@@ -139,7 +132,7 @@ class OsdScrub {
    * would not be retried before 'delay' seconds have passed.
    */
   void delay_on_failure(
-      Scrub::ScrubJobRef sjob,
+      Scrub::ScrubJob& sjob,
```

Comment:
This doesn't obviously make sense to me.  There appears to be only one caller:

```
/*
 * note: arbitrary delay used in this early version of the
 * scheduler refactoring.
 */
void PgScrubber::penalize_next_scrub(Scrub::delay_cause_t cause)
{
  m_osds->get_scrub_services().delay_on_failure(
      *m_scrub_job, 5s, cause, ceph_clock_now());
}
```

The implementation

```
void OsdScrub::delay_on_failure(
      Scrub::ScrubJob& sjob,
      std::chrono::seconds delay,
      Scrub::delay_cause_t delay_cause,
      utime_t now_is)
{
  m_queue.delay_on_failure(sjob, delay, delay_cause, now_is);
}
```

simply invokes ScrubQueue::delay_on_failure

```
void ScrubQueue::delay_on_failure(
    Scrub::ScrubJob& sjob,
    std::chrono::seconds delay,
    Scrub::delay_cause_t delay_cause,
    utime_t now_is)
{
  dout(10) << fmt::format(
		  "pg[{}] delay_on_failure: delay:{} now:{:s}",
		  sjob.pgid, delay, now_is)
	   << dendl;
  sjob.delay_on_failure(delay, delay_cause, now_is);
}
```

which finally invokes ScrubJob::delay_on_failure.  This seems like a holdover from the previous scheme.  PgScrubber::requeue_penalized, by contrast, seems to be perfectly happy invoking ScrubJob::delay_on_failure directly:

```
void PgScrubber::requeue_penalized(Scrub::delay_cause_t cause)
{
  /// \todo fix the 5s' to use a cause-specific delay parameter
  m_scrub_job->delay_on_failure(5s, cause, ceph_clock_now());
  m_osds->get_scrub_services().enqueue_target(*m_scrub_job);
}
```

Let's remove OsdScrub::delay_on_failure and ScrubQueue::delay_on_failure.

Reply:
Ah, and PgScrubber::penalize_next_scrub() actually has no further callers since PgScrubber::requeue_penalized calls ScrubJob::delay_on_failure and OsdScrub::enqueue_target.

Reply:
Yes - the whole chain was irrelevant after the changes. Removed.